src/cluster-client/src/metrics.rs

```diff
@@ -0,0 +1,134 @@
+                help: "The total sum of dataflow wallclock lag measurements.",
+                var_labels: ["instance_id", "replica_id", "collection_id"],
+            )),
+            dataflow_wallclock_lag_seconds_count: metrics_registry.register(metric!(
+                name: "mz_dataflow_wallclock_lag_seconds_count",
+                help: "The total count of dataflow wallclock lag measurements.",
+                var_labels: ["instance_id", "replica_id", "collection_id"],
+            )),
+        }
+    }
+
+    /// Return an object that tracks wallclock lag metrics for the given collection on the given
+    /// cluster and replica.
+    pub fn wallclock_lag_metrics(
+        &self,
+        collection_id: String,
+        instance_id: Option<String>,
+        replica_id: Option<String>,
+    ) -> WallclockLagMetrics {
+        let labels = vec![
+            instance_id.unwrap_or_default(),
+            replica_id.unwrap_or_default(),
+            collection_id.to_string(),
+        ];
+
+        let labels_with_quantile = |quantile: &str| {
+            labels
+                .iter()
+                .cloned()
+                .chain([quantile.to_string()])
+                .collect()
+        };
+
+        let wallclock_lag_seconds_min = self
+            .dataflow_wallclock_lag_seconds
+            .get_delete_on_drop_metric(labels_with_quantile("0"));
+        let wallclock_lag_seconds_max = self
+            .dataflow_wallclock_lag_seconds
+            .get_delete_on_drop_metric(labels_with_quantile("1"));
+        let wallclock_lag_seconds_sum = self
+            .dataflow_wallclock_lag_seconds_sum
+            .get_delete_on_drop_metric(labels.clone());
+        let wallclock_lag_seconds_count = self
+            .dataflow_wallclock_lag_seconds_count
+            .get_delete_on_drop_metric(labels);
+        let wallclock_lag_minmax = SlidingMinMax::new(60);
+
+        WallclockLagMetrics {
+            wallclock_lag_seconds_min,
+            wallclock_lag_seconds_max,
+            wallclock_lag_seconds_sum,
+            wallclock_lag_seconds_count,
+            wallclock_lag_minmax,
+        }
+    }
+}
+
+/// Metrics tracking frontier wallclock lag for a collection.
+#[derive(Debug)]
+pub struct WallclockLagMetrics {
+    /// Gauge tracking minimum dataflow wallclock lag.
+    wallclock_lag_seconds_min: DeleteOnDropGauge<'static, AtomicF64, Vec<String>>,
+    /// Gauge tracking maximum dataflow wallclock lag.
+    wallclock_lag_seconds_max: DeleteOnDropGauge<'static, AtomicF64, Vec<String>>,
+    /// Counter tracking the total sum of dataflow wallclock lag.
+    wallclock_lag_seconds_sum: DeleteOnDropCounter<'static, AtomicF64, Vec<String>>,
+    /// Counter tracking the total count of dataflow wallclock lag measurements.
+    wallclock_lag_seconds_count: DeleteOnDropCounter<'static, AtomicU64, Vec<String>>,
+
+    /// State maintaining minimum and maximum wallclock lag.
+    wallclock_lag_minmax: SlidingMinMax<f32>,
```

Comment:
Also a bit confused here, if we're storing F64 above why do we use f32 here?

Reply:
To safe memory! There is some more discussion in the comments [here](https://github.com/MaterializeInc/materialize/pull/29140), but the gist is that it reduces the memory required for tracking this metric from ~1.5KB to half that, for each collection. There can be a lot of collections in an environment, so the difference matters.